data/source/MongoDb.php

```diff
@@ -212,91 +233,98 @@ public static function enabled($feature = null) {
 	 *          before timing out and throwing an exception. Defaults to `100`.
 	 *        - `'schema'` _\Closure_: A closure or anonymous function which returns the schema
 	 *          information for a model class. See the `$_schema` property for more information.
-	 *        - `'gridPrefix'` _string_: The default prefix for MongoDB's `chunks` and `files`
-	 *          collections. Defaults to `'fs'`.
-	 *        - `'replicaSet'` _string_: See the documentation for `Mongo::__construct()`. Defaults
-	 *          to `false`.
-	 *        - `'readPreference'` _mixed_: May either be a single value such as Mongo::RP_NEAREST,
-	 *          or an array containing a read preference and a tag set such as:
-	 *          (Mongo::RP_SECONDARY_PREFERRED, ['dc' => 'east] See the documentation for
-	 *          `Mongo::setReadPreference()`. Defaults to null.
-	 *          Typically, these parameters are set in `Connections::add()`, when adding the
-	 *          adapter to the list of active connections.
 	 *
 	 *        Disables auto-connect, which is by default enabled in `Source`. Instead before
 	 *        each query execution the connection is checked and if needed (re-)established.
 	 * @return void
 	 */
 	public function __construct(array $config = []) {
-		if (class_exists($server = $this->_classes['server'], false)) {
-			$defaultHost = $server::DEFAULT_HOST . ':' . $server::DEFAULT_PORT;
-		} else {
-			$defaultHost = static::DEFAULT_HOST . ':' . static::DEFAULT_PORT;
-		}
 		$defaults = [
-			'host' => $defaultHost,
-			'login'      => null,
-			'password'   => null,
-			'database'   => null,
-			'timeout'    => 100,
-			'replicaSet' => false,
-			'schema'     => null,
-			'gridPrefix' => 'fs',
-			'w'          => 1,
-			'wTimeoutMS' => 10000,
-			'readPreference' => null,
-			'autoConnect' => false,
-			'dsn' => null
+			'host'               => static::DEFAULT_HOST . ':' . static::DEFAULT_PORT,
+			'login'              => null,
+			'password'           => null,
+			'database'           => null,
+			'dsn'                => null,
+			'timeout'            => 100,
+			'uriOptions'            => [
+				'w'                        => 'majority',
+				'wTimeoutMS'               => 10000,
+				'journal'                  => true,
+				'readConcernLevel'         => 'local',
+				'readPreference'           => 'primary',
+				'readPreferenceTags'       => [],
+			],
+			'driverOptions' => [
+				'allow_invalid_hostname' => false,
+				'weak_cert_validation'   => false
+			]
 		];
-		parent::__construct($config + $defaults);
+		$config = Set::merge($defaults, $config);
+		if (!isset($config['uriOptions']['connectTimeoutMS']) && $config['timeout']) {
+			$config['uriOptions']['connectTimeoutMS'] = $config['timeout'];
+		}
+		parent::__construct($config);
 	}
 
 	/**
 	 * Initializer. Adds operator handlers which will later allow to correctly cast any
-	 * values. Constructs a DSN from configuration.
+	 * values. Constructs a DSN from configuration, if not given.
```

Comment:
We'll get back to DSN handling over all data sources in another changeset. This is good enough now :)

Reply:
I took care of DSN Handling in https://github.com/UnionOfRAD/lithium/commit/85deea943b88392702c895ad3bad233935c36fb7. My goal was to support connection strings, displayed by MongoDB Atlas without any changes.

I wrote test cases for that and implemented accordingly. I thought about adding/expanding the HostString class (or create a new `DsnString`). If you want me to take care of that, I already have a branch for that.

Reply:
That would be great to have consistent DSN support across all the data source adapters! But let's get this PR merged first. I don't want to hold it back with anything.